### PR TITLE
ocl-icd 2.3.1 (new formula)

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Split off from #103141

See other PR for why this we need to use this instead of `opencl-icd-loader`.

Also, this is the option most major Linux distros have picked (https://repology.org/project/ocl-icd/versions) like Debian/Ubuntu, Fedora, Arch, Gentoo, ...

### CODE REVIEW

File: Formula/ocl-icd.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  uses_from_macos "libxml2" => :build
+  uses_from_macos "libxslt" => :build
+  uses_from_macos "ruby" => :build
+
+  def install
+    ENV["XML_CATALOG_FILES"] = etc/"xml/catalog"
+    system "./bootstrap"
+    system "./configure", *std_configure_args,
+                          "--disable-silent-rules",
+                          "--enable-custom-vendordir=#{etc}/OpenCL/vendors"
```
Comment (cho-m) [inline Formula/ocl-icd.rb:27]: As I mentioned in #103141:

There is a minor question on whether `etc` is best location. The ICD files need to be updated by corresponding formulae since they contain Cellar path. I think this does happen right now when file is installed by `cmake` as opposed to `etc.install`, though I don't know if that is intended.

File: Formula/ocl-icd.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  def caveats
+    s = "The default vendors directory is #{etc}/OpenCL/vendors\n"
+    on_linux do
+      s += <<~EOS
+        No OpenCL implementation is pre-installed, so all dependents will require either
+        installing a compatible formula or creating an ".icd" file mapping to a manually
+        installed option. A portable implementation is available via the `pocl` formula.
+      EOS
+    end
```
Comment (cho-m) [inline Formula/ocl-icd.rb:40]: Officially, the specification explicitly says to use `/etc/OpenCL/vendors`, which is why I added first caveat.

The latter part is due to the fact that the intended distribution of OpenCL stack differs from other formulae where I haven't pre-installed an OpenCL implementation/runtime. On macOS, the OpenCL framework plays this role.

Alternatively, we could force install `pocl` to get runtime functionality; however, this may not be useful depending on user's hardware as Intel/AMD/Nvidia implementations are better tuned to specific hardware. At least some Intel iGPU & AMD GPU have open-source options. Nvidia is mainly proprietary.

Comment (danielnachun) [reply Formula/ocl-icd.rb:40]: As I understand it, if we install `pocl` this doesn't preclude the user installing other OpenCL backends and using those instead.  They just have to configure everything so that their backend is used instead of `pocl`.  If that is correct, then I think we can just always install `pocl` and update the caveat to say that we have a shipped a very basic OpenCL backend for testing purposes but the user should install a higher performance backend tuned to their specific hardware for more intensive applications.

The Intel and AMD implementations are open source and do appear compatible with Homebrew (possibly aside from notability requirements), but they could be difficult to test in a CI environment, so it may not be feasible for us to ship them.  

As long as our caveat makes it clear that the user will need to take further action to get a performant OpenCL implementation, I think our approach to this will be acceptable.

File: Formula/ocl-icd.rb (hunk 1 above)
Comment (danielnachun) [reply Formula/ocl-icd.rb:27]: Is the concern here that once the `icd` file is installed the first time, it won't be properly overridden in subsequent updates and would contain an outdated Cellar path?  I think `etc` seems like the right place to put the files but I'm not sure how we update files in there once they've been installed.  I know we don't delete files from `etc` upon uninstall, but I'm not sure if we can overwrite files there.

**Comment by cho-m** [comment]
> During the testing things did not work quite as well. `clinfo`, `clfft` and `hashcat` all threw some sort of error because they couldn't find an OpenCL device. I'm not yet sure if this because they don't know to search `HOMEBREW_PREFIX/etc/OpenCL/vendors` or because of some other issue with `pocl`. I'll be looking into this more but @cho-m let me know if you have any other thoughts.

Not sure. I was able to use `clinfo` and `hashcat` in Linux Docker container with just having dependencies.

It was even possible to build `clinfo` without adding dependencies to formula and get it to run (due to `HOMEBREW_PREFIX/lib` in RPATH). In this example, you can see it fail once `ocl-icd` is unlinked:
```console
$ clinfo
Number of platforms                               1
  Platform Name                                   Portable Computing Language
  Platform Vendor                                 The pocl project
  Platform Version                                OpenCL 3.0 PoCL 3.0-rc2  Linux, Release, RELOC, LLVM 14.0.6, SLEEF, DISTRO, POCL_DEBUG
...
ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.3.1
  ICD loader Profile                              OpenCL 3.0

$ brew unlink ocl-icd
Unlinking /home/linuxbrew/.linuxbrew/Cellar/ocl-icd/2.3.1... 10 symlinks removed.

$ clinfo
clinfo: error while loading shared libraries: libOpenCL.so.1: cannot open shared object file: No such file or directory
```

File: Formula/ocl-icd.rb (hunk 1 above)
Comment (cho-m) [reply Formula/ocl-icd.rb:27]: > Is the concern here that once the `icd` file is installed the first time, it won't be properly overridden in subsequent updates and would contain an outdated Cellar path?

Yes, that is main worry as seen by what is inside the file:
```console
$ cat ~/.linuxbrew/etc/OpenCL/vendors/pocl.icd; echo
/home/linuxbrew/.linuxbrew/Cellar/pocl/3.0/lib/libpocl.so.2.9.0
```

---

> I think `etc` seems like the right place to put the files but I'm not sure how we update files in there once they've been installed. I know we don't delete files from `etc` upon uninstall, but I'm not sure if we can overwrite files there.

When I tried updating from PoCL 1.8 to 3.0, I did see that `cmake` overwrote the previous file. I don't know if this would behave differently if I modified `pocl.icd`

However, I am not familiar enough with brew `etc` logic to know if this is expected.

Comment (danielnachun) [reply Formula/ocl-icd.rb:27]: @Homebrew/core do files in `etc` get overwritten during formula updates if they've been modified by the user?  If not we probably need to change the cellar path to the opt dir and use the unversioned soname, as that combination should still work if the file is not updated after a version update.

**Comment by danielnachun** [comment]
> Not sure. I was able to use `clinfo` and `hashcat` in Linux Docker container with just having dependencies.
> 
> It was even possible to build `clinfo` without adding dependencies to formula and get it to run (due to `HOMEBREW_PREFIX/lib` in RPATH). In this example, you can see it fail once `ocl-icd` is unlinked:
> 
> ```
> $ clinfo
> Number of platforms                               1
>   Platform Name                                   Portable Computing Language
>   Platform Vendor                                 The pocl project
>   Platform Version                                OpenCL 3.0 PoCL 3.0-rc2  Linux, Release, RELOC, LLVM 14.0.6, SLEEF, DISTRO, POCL_DEBUG
> ...
> ICD loader properties
>   ICD loader Name                                 OpenCL ICD Loader
>   ICD loader Vendor                               OCL Icd free software
>   ICD loader Version                              2.3.1
>   ICD loader Profile                              OpenCL 3.0
> 
> $ brew unlink ocl-icd
> Unlinking /home/linuxbrew/.linuxbrew/Cellar/ocl-icd/2.3.1... 10 symlinks removed.
> 
> $ clinfo
> clinfo: error while loading shared libraries: libOpenCL.so.1: cannot open shared object file: No such file or directory
> ```

Okay that's very helpful to know.  I'll double check this and see if I can get `clinfo`, `hashcat` and `clfft` to work.  If there's no issues with them I think we will be very close to merging this!

File: Formula/ocl-icd.rb (hunk 1 above)
Comment (carlocab) [reply Formula/ocl-icd.rb:27]: `etc.install` will not overwrite files in `etc`. Practically speaking, this means that files in `etc` are never overwritten when installing from bottles.

You may find that building from source overwrites files in `etc`, though. (But that's because `make`, etc did it, and not `brew`)

Comment (danielnachun) [reply Formula/ocl-icd.rb:27]: In that case, I think should `inreplace` the Cellar path with the opt dir path and the unversioned soname when installing to `etc`.

Comment (carlocab) [reply Formula/ocl-icd.rb:27]: Is `pocl.icd` meant to be user-configurable, or do you just want a persistent `OpenCL/Vendors` directory for users to stick their other `*.icd` files into?

If it's the former, we can do an `inreplace`. If it's the latter we might want to just create a symlink from `etc/"OpenCL/Vendors/pocl.icd"` to a `pocl.icd` in the appropriate Cellar/opt path.

Comment (danielnachun) [reply Formula/ocl-icd.rb:27]: We should see what @cho-m thinks but my understanding is that the we just want a persistent `OpenCL/vendors` directory in `etc` for users to add `*.icd`.  The only content in `pocl.icd` is the path to the `pocl` shared library, which is not something users should be changing anyway.  So the symlinking strategy seems to make more sense to me at the moment.

**Comment by danielnachun** [comment]
I was able to get `hashcat` and `clinfo` to work after upgrading the `pocl` formula to 3.0.  The test binary for `clfft` still segfaults, possibly because it still isn't detecting an OpenCL device.  However I think there is enough evidence that this is working as expected that I'm ready to approve this once we resolve the `etc` issue.

File: Formula/ocl-icd.rb (hunk 1 above)
Comment (cho-m) [reply Formula/ocl-icd.rb:27]: Not sure if there is a common use-case for modifying content of a formula-installed ICD file.

There are some scenarios where users may want to rename file to temporarily disable an OpenCL runtime without uninstalling (as files without `.icd` are ignored), but that should still be possible with symlinks.

---

Either approach would need to be handled in the corresponding formulae like `pocl`, `oclgrind`, ..., so current `etc` directory sounds like it will work.

Comment (danielnachun) [reply Formula/ocl-icd.rb:27]: That makes sense and should be easy enough to set up in those respective formulae.  I think we've got this figured out.

File: Formula/ocl-icd.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    pkgshare.install "ocl_test.c"
+  end
+
+  def caveats
+    s = "The default vendors directory is #{etc}/OpenCL/vendors\n"
+    on_linux do
+      s += <<~EOS
+        No OpenCL implementation is pre-installed, so all dependents will require either
+        installing a compatible formula or creating an ".icd" file mapping to a manually
+        installed option. A portable implementation is available via the `pocl` formula.
```
Comment (danielnachun) [inline Formula/ocl-icd.rb:38]: Let's rephrase this to:
```
No OpenCL implementation is pre-installed, so all dependents will require either
installing a compatible formula or creating an ".icd" file mapping to an externally
installed implementation. Any ".icd" files copied or symlinked into 
`#{etc}/OpenCL/vendors` will automatically be detected by `ocl-icd`. A portable
OpenCL implementation is available via the `pocl` formula.
```
Please fix anything that may be incorrect there!

**Review by danielnachun — APPROVED** [review]
I think this is ready to publish now.
